staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go

```diff
@@ -0,0 +1,194 @@
+package filters
+
+import (
+	"context"
+	"fmt"
+	"github.com/stretchr/testify/assert"
+	"io/ioutil"
+	fctypesv1a1 "k8s.io/api/flowcontrol/v1alpha1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/authentication/user"
+	apifilters "k8s.io/apiserver/pkg/endpoints/filters"
+	epmetrics "k8s.io/apiserver/pkg/endpoints/metrics"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/server/mux"
+	utilflowcontrol "k8s.io/apiserver/pkg/util/flowcontrol"
+	fq "k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing"
+	fcmetrics "k8s.io/apiserver/pkg/util/flowcontrol/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+	"k8s.io/component-base/metrics/testutil"
+	"k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap"
+	"net"
+	"net/http"
+	"testing"
+	"time"
+)
+
+const (
+	decisionExecute = iota
+	decisionCancel
+)
+
+type fakeApfFilter struct {
+	mockDecision int
+}
+
+func (t fakeApfFilter) Handle(ctx context.Context,
+	requestDigest utilflowcontrol.RequestDigest,
+	noteFn func(fs *fctypesv1a1.FlowSchema, pl *fctypesv1a1.PriorityLevelConfiguration),
+	queueNoteFn fq.QueueNoteFn,
+	execFn func(),
+) {
+	noteFn(bootstrap.SuggestedFlowSchemaGlobalDefault, bootstrap.SuggestedPriorityLevelConfigurationGlobalDefault)
+	switch t.mockDecision {
+	case decisionExecute:
+		queueNoteFn(true)
+		execFn()
+	case decisionCancel:
+		queueNoteFn(true)
+	}
+}
+
+func (t fakeApfFilter) Run(stopCh <-chan struct{}) error {
+	return nil
+}
+
+func (t fakeApfFilter) Install(c *mux.PathRecorderMux) {
+}
+
+func newApfDecisionHandler(decision int) http.Handler {
+	longRunningRequestCheck := BasicLongRunningRequestCheck(sets.NewString("watch"), sets.NewString("proxy"))
+	apfHandler := WithPriorityAndFairness(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+	}), longRunningRequestCheck, fakeApfFilter{mockDecision: decision})
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		r = r.WithContext(apirequest.WithUser(r.Context(), &user.DefaultInfo{
+			Groups: []string{"system:unauthenticated"},
+		}))
+		apfHandler.ServeHTTP(w, r)
+	})
+}
+
+const apfServerAddress = "127.0.0.1:6443"
+
+func newApfServer(apfHandler http.Handler) (*http.Server, error) {
```

Comment:
i presume this should be replace by `httptest.NewServer` which doesnt necessarily bind a constant address/port in the test. otherwise the test will fail on running in parallel, isnt it?

Reply:
Fixed, and thanks for the explanation!